### ISSUE TITLE
5.0 Beta bug: Pihole -g doesn't delete duplicates

### ISSUE THREAD

**Issue by p1r473** [pi-hole/pi-hole#3286]
**In raising this issue, I confirm the following:** `{please fill the checkboxes, e.g: [X]}`

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] The issue I am reporting can be *replicated*.
- [X] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/pi-hole/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/pi-hole/issues)).

**How familiar are you with the the source code relevant to this issue?:**

0

---
**Expected behaviour:**

Gravity.db contains no duplicates

**Actual behaviour:**

Gravity.db contains duplicates. This causes the database to balloon in size.

**Steps to reproduce:**

Use only 2 identical adlists that contain same hosts. For example, any of:
https://any.andryou.com/oisddbl.txt
https://raw.githubusercontent.com/ookangzheng/dbl-oisd-nl/master/dbl.txt
https://dbl.oisd.nl/
https://cdn.andryou.com/oisddbl.txt
https://gitlab.com/ookangzheng/dbl-oisd-nl/raw/master/dbl.txt

All of these lists have the same duplicate 1109764 hosts and both hostsfiles are identical. In the end, I should have 1109764 domains. However, I have 2219528 domains, which is 2x the amount of domains.

**Debug token provided by [uploading `pihole -d` log](https://discourse.pi-hole.net/t/the-pihole-command-with-examples/738#debug):**

https://tricorder.pi-hole.net/iknrrxtfbf

**Troubleshooting undertaken, and/or other relevant information:**

This is a problem in the backend code and cannot be fixed on the user's end.

![image](https://user-images.githubusercontent.com/9235633/79704330-e81e2000-827e-11ea-8501-621c4ccc0d1b.png)

Every entry is listed twice:

![image](https://user-images.githubusercontent.com/9235633/79704629-439cdd80-8280-11ea-9beb-0d6f86a36363.png)

**Comment by DL6ER** [comment]
As you can see, the domains you see are not full duplicates but differ in the `adlist_id`. Image you have two ad lists with the following contents (this is an academic example, but will work):
```
# Child 1 "Anna"
youtube.com
facebook.com
```
```
# Child 2 "Peter"
twitter.com
facebook.com
```

This will lead to gravity containing:
```
youtube.com  | 1
facebook.com | 1
twitter.com  | 2
facebook.com | 2
```

As you can see, the two lists lead to `facebook.com` being included two times. However, imagine now you configure the smartphone of Anna to use *only* list 1 and the smartphone of Peter to use *only* list 2. Anna will not be able to access youtube.com, whereas Peter will not able to access twitter.com. However, both get facebook.com blocked.

If we would not store the duplicates, Pi-hole could not decide whether a device with only a subset of the available ad lists should get a certain page blocked or not. We have carefully investigated a number of different approached, including an approach to consolidate list entries like
```
youtube.com  | 1
facebook.com | 1,2
twitter.com  | 2
```
to avoid duplicates. However, this turned out to be **much** less performant compared to our current implementation. In the end, we decided to pay exponential performance-gains (orders of magnitude) by linear increase of database file size (linearly proportional to the number of lists).

**TL;DR** This is not a bug, but, in fact, even completely intended and consequence of the per-client adlist ability.

The fact that you see a ton of duplicates is not a gravity issue but rather a list issue: If they are basically identical (contain numerous duplicates), you can maybe just get rid of one of them.

**Comment by p1r473** [comment]
Is there a way to disable the per-client adlist so I can shrink my database?

**Comment by DL6ER** [comment]
Only consolidating the lists beforehand so Pi-hole sees them as one list.

**Comment by p1r473** [comment]
Ok, so I have a few options I guess.

Try and learn SQL and figure out how to remove duplicates myself (if you have a command I can run that would be great)
Or
Create my own adlist downloader. 

I'd love to have the ability to disable this functionality, maybe a feature request, as I live alone, I only need 1 different blocklist, and my database is now 1gb

Thanks @DL6ER

**Comment by DL6ER** [comment]
In case you want to remove duplicates, run
```
sqlite3 /etc/pihole/gravity.db "DELETE FROM gravity WHERE rowid NOT IN (SELECT rowid FROM gravity GROUP BY domain); VACUUM;"
```
This ensures that each domain in only once in the database (it may take a long time to remove the duplicates if your database is very large).

Just as a side note: My database is roughly 6 MB with the stock lists, you may want to rethink the amount of sources you're using.

**Comment by PromoFaux** [comment]
@p1r473 You're downloading two identical lists, I would just disable one of them.

The `adlist` table is where the contents of each adlist are stored.. similar to how they are stored in the current flat file system (e.g each list is stored in `/etc/pihole/`, as well as the combined `gravity.list`)

@DL6ER is correct in that this is also how the per-client stuff works, also, but I see your query specifically

**Comment by p1r473** [comment]
@DL6ER actually I only enabled the 2 above lists for the simplified pihole -g out. My actual 450 sources made a 1gb list.

Thanks for the SQL command! I'll put it in a cronjob. I'd still love a feature whereby I can disable this per-client adlist and have dupes removed.
@PromoFaux Many of my lists have the same entries

![image](https://user-images.githubusercontent.com/9235633/79702306-e7cb5800-8271-11ea-9267-5da7d71dbdcb.png)

**Comment by PromoFaux** [comment]
@p1r473 I mistakenly commented too early instead of deleting, I got myself confused. That said, as I see it here...

In 4.x, when gravity downloads lists it stores them in `/etc/pihole` as, for example, `list.0.raw.githubusercontent.com.domains`

These are then combined and deduplicated into `list.preEventHorizon`, and then the user whitelists/blacklists are applied and `gravity.list` is created.

So the behaviour isn't all that different, perhaps... though I may understand it wrong. Do you still have a 4.x install running? I would be curious to see the size of the `/etc/pihole` directory with all 47 of the lists you are using...

**Comment by p1r473** [comment]
@PromoFaux are you saying we shoulnd't be having duplicates, as @DL6ER said we should be

**Comment by PromoFaux** [comment]
No, I'm saying that it shouldn't be too dissimilar in size to the current (4.x) way of storing things.

simplified example of what I am saying for 4.x:

`https://blah.com/list.txt contains` 1 entry "www.google.com"
`https://another.com/list.txt contains` 1 entry "www.google.com"

when gravity runs, those lists are stored as two files (the cached files to prevent redownloading lists that haven't changed):
`/etc/pihole/list.0.blah.com.domains`
`/etc/pihole/list.1.another.com.domains`

At this stage "www.google.com" is stored twice (albeit in two separate files)

A deduplicated copy of that list is created under:
`/etc/pihole/list.preEventHorizon`

At this stage "www.google.com" is stored three times.

Finally the blacklist/whitelist are added/subtracted from this file (assuming nothing in either of those), the final file is created:
`/etc/pihole/gravity.list`

"www.google.com" is now stored 4 times

@DL6ER is correct, with how gravity database is designed, there should be duplicates.

**Comment by p1r473** [comment]
Thanks guys
Well I will implement the above SQL dedupe command in a cronjob weekly
I'll close this issue but, feature request- ability to turn off the per-client adlist so there are no dupes

**Comment by PromoFaux** [comment]
@p1r473 , out of morbid curiosity, please could you tell me all 47 of your lists?

Thanks

**Comment by p1r473** [comment]
It is for sure overkill :)
https://pastebin.com/raw/QbyzvFNi

**Comment by PromoFaux** [comment]
Whew, OK that took a long time to run.. but confirmed 1.4GB size. 

I had some thoughts on de-duping to optimise in the gravity process, but even if anything comes from them, it will not be in time for 5.0, so I'll keep them to myself for now.

The only thing I can think that would speed up/improve it from your own point of view is if you were to download and merge all of those lists into a single file, and add that to the adlists section. You will, of course, lose the ability to query which adlist a domain is found in. Likewise if you run any deduplication against the database as it stands, a result from `pihole -q` will only give you the name of _one_ list that it is contained in, vs all of the lists if you left the database as is.

**Comment by PromoFaux** [comment]
Or, go easy on the number of domains blocked.. bigger != better... but ultimately, that comes down to preference, and I'm not going to tell you specifically _not_ to add so many lists/domains... just keep in mind that we can only support so much...

**Comment by p1r473** [comment]
> In case you want to remove duplicates, run
> 
> ```
> sudo sqlite3 /etc/pihole/gravity.db "DELETE FROM gravity WHERE rowid NOT IN (SELECT rowid FROM gravity GROUP BY domain); VACUUM;"
> ```
> 
> This ensures that each domain in only once in the database (it may take a long time to remove the duplicates if your database is very large).
> 
> Just as a side note: My database is roughly 6 MB with the stock lists, you may want to rethink the amount of sources you're using.

Hey @DL6ER do I need to also clean the domainlist table, or just the gravity table?

**Comment by p1r473** [comment]
Was able to shrink from 1.4gb to 200 megs.
Thanks for the help!

**Comment by p1r473** [comment]
> Whew, OK that took a long time to run.. but confirmed 1.4GB size.
> 
> I had some thoughts on de-duping to optimise in the gravity process, but even if anything comes from them, it will not be in time for 5.0, so I'll keep them to myself for now.
> 
> The only thing I can think that would speed up/improve it from your own point of view is if you were to download and merge all of those lists into a single file, and add that to the adlists section. You will, of course, lose the ability to query which adlist a domain is found in. Likewise if you run any deduplication against the database as it stands, a result from `pihole -q` will only give you the name of _one_ list that it is contained in, vs all of the lists if you left the database as is.

@PromoFaux  there is one interesting part. While the combined database is 1.4GB, all of the .lists files add up to only 417MB, meaning the sum is greater than the parts. Any idea how the database becomes more than 3x bigger than the individual lists?

**Comment by PromoFaux** [comment]
SQL databases have overheads, which will use additional storage space, for things like indexes which make them vastly superior for querying data over how we were doing it with manipulating text files. 

Because of the database, and the way things are being stored, we are able to do things with Pi-hole that we could only dream of before now.

**Comment by DL6ER** [comment]
I'm not against optimizing the database, however, I do ask myself if this is really an optimization. As I already said, we traded performance by (disk file) size but there is another dimension to take into account here: Cache size. Currently, FTL has to hold one index in memory to rapidly reply to whether a domain is in gravity or not. With a linked list, this would double.

The additional effort for four identical domains from different adlists in the index is minimal:
```
domain (hash) | adlistID | database pointer (on-disk)
0xeff43e67    | 1        | 0xeff980ce
0xeff43e67    | 2        | 0x41bbdac6
0xeff43e67    | 3        | 0xd9459755
0xeff43e67    | 4        | 0x3b143d40
```
We do not really care about the content of the `gravity` table once the index has been generated. We could even detach the index from the table and flush it completely if this is possible. However, this would leave the database in some dangling intermediate state which would be hard to understand for users trying to evaluate their own SQL statements against the database. So I strongly vote against this. Note that, whatever we do, will take substantially longer during the building of the database.
Not sure what @PromoFaux observed, however, if we observe
```sql
INSERT INTO gravity (domain,adlistID) VALUES ('somedomain.com', 1);
```
by something like

```sql
INSERT OR IGNORE INTO gravity_domains (domain) VALUES ('somedomain.com');
INSERT INTO gravity (domain, adlistID) VALUES ((SELECT id FROM gravity_domains WHERE domain = 'somedomain.com'), 1);
```
would create a lot of overhead as instead of writing one after another to disk, it always involves  a lookup in a (necessarily!) growing index which needs to be rebuilt after each domain insertion into `gravity_domains`. Note that we intentionally delay the index generation of the `gravity` table until after the table has been created completely.

Some of the additional delay may be compensated if you ran your tests on a beefy machine with enough RAM, however, we should not expect this to the the standard environment Pi-hole is run on.

**Comment by p1r473** [comment]
For now, I am just cleaning up everything myself in a simple script (reposted from the other issue for anyone else who wanders to this issue:)
```
#Delete disabled adlists:
sudo sqlite3 /etc/pihole/gravity.db "DELETE FROM adlist WHERE enabled = 0;"
#Delete disabled whitelist/blacklist
sudo sqlite3 /etc/pihole/gravity.db "DELETE FROM domainlist WHERE enabled = 0;"
#Delete duplicates from gravity
sudo sqlite3 /etc/pihole/gravity.db "DELETE FROM gravity WHERE rowid NOT IN (SELECT rowid FROM gravity GROUP BY domain);"
#Vacuum the DB
sudo sqlite3 /etc/pihole/gravity.db "VACUUM;"
```

**Comment by p1r473** [comment]
> For now, I am just cleaning up everything myself in a simple script (reposted from the other issue for anyone else who wanders to this issue:)
> 
> ```
> #Delete disabled adlists:
> sudo sqlite3 /etc/pihole/gravity.db "DELETE FROM adlist WHERE enabled = 0;"
> #Delete disabled whitelist/blacklist domains
> sudo sqlite3 /etc/pihole/gravity.db "DELETE FROM domainlist WHERE enabled = 0;"
> #Delete duplicates from gravity
> sudo sqlite3 /etc/pihole/gravity.db "DELETE FROM gravity WHERE rowid NOT IN (SELECT rowid FROM gravity GROUP BY domain);"
> #Vacuum the DB
> sudo sqlite3 /etc/pihole/gravity.db "VACUUM;"
> ```

Nope, this plan of attack failed. Issue persists. Installing PHP 7.4 and all the generic packages, then removing (not purging) all the 7,0 packages, produces the same result.

@DL6ER I am back into the bugged state. What would you like me to test for the intl package?
